subsys/storage/flash_map/flash_map.c

```diff
@@ -262,3 +269,60 @@ struct device *flash_area_get_device(const struct flash_area *fa)
 {
 	return device_get_binding(fa->fa_dev_name);
 }
+
+#if defined(CONFIG_FLASH_AREA_CHECK_INTEGRITY)
+int flash_area_check_int_sha256(const struct flash_area *fa,
+				const struct flash_area_check *fac)
+{
+	unsigned char hash[TC_SHA256_DIGEST_SIZE];
+	struct tc_sha256_state_struct sha;
+	struct device *dev;
+	int to_read;
+	int pos;
+	int rc;
+
+	if (!fa || !fac || !fac->match || !fac->rbuf ||
+	    fac->clen <= 0 || fac->off < 0 || fac->rblen <= 0) {
+		return -EINVAL;
+	}
+
+	if (!is_in_flash_area_bounds(fa, fac->off, fac->clen)) {
+		return -EINVAL;
+	}
+
+	if (tc_sha256_init(&sha) != TC_CRYPTO_SUCCESS) {
+		return -EIO;
+	}
+
+	dev = device_get_binding(fa->fa_dev_name);
+	to_read = fac->rblen;
+
+	for (pos = 0; pos < fac->clen; pos += to_read) {
+		if (pos + to_read > fac->clen) {
+			to_read = fac->clen - pos;
+		}
+
+		rc = flash_read(dev, (fa->fa_off + fac->off + pos),
+				fac->rbuf, to_read);
+		if (rc != 0) {
+			return rc;
+		}
+
+		if (tc_sha256_update(&sha,
+				     fac->rbuf,
+				     to_read) != TC_CRYPTO_SUCCESS) {
+			return -EIO;
+		}
```

Comment:
I ma not sure about the returned error codes. For example `espi_xec_flash_read` will return `-EIO` when flash is disabled same thing happens for all the sha functions that fail here. 
You call the function and get the error, now what you do? You do not know whether the sha was bad and you need to fix the flash contents or the flash is bad which means that fixing flash contents will not work.

Reply:
Agreed! Thanks for point me that. My suggestion is use ESRCH _(No such context)_ for SHA operations and EILSEQ _( Illegal byte sequence )_ for wrong match.